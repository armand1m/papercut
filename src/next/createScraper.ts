import { createLogger } from './createLogger';
import { createRunner } from './createRunner';

export interface ScraperProps {
  name: string;
  options?: Partial<ScraperOptions>;
}

export interface ScraperOptions {
  log: boolean;
  cache: boolean;
  concurrency: {
    page: number;
    node: number;
    selector: number;
  };
}

export interface PaginationOptions {
  createPaginatedUrl: (
    metadata: ScraperProps,
    pageNumber: number
  ) => string;
  lastPageNumberSelector: string;
}

export const defaultOptions: ScraperOptions = {
  log: process.env.DEBUG === 'true',
  cache: false,
  concurrency: {
    page: 2,
    node: 2,
    selector: 2,
  },
};

export type Scraper = ReturnType<typeof createScraper>;

export const createScraper = (props: ScraperProps) => {
  const options: ScraperOptions = {
    ...defaultOptions,
    ...props.options,
    concurrency: {
      ...defaultOptions.concurrency,
      ...props.options?.concurrency,
    },
  };

  const logger = createLogger({
    name: props.name,
    enabled: options.log,
  });

  return {
    run: createRunner({
      logger,
      options,
    }),
  };
};
