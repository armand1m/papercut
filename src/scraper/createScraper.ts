import { createLogger } from './createLogger';
import { createRunner } from './createRunner';

export interface ScraperProps {
  /**
   * The scraper name.
   * This will be used only for logging purposes.
   */
  name: string;
  /**
   * The scraper options.
   * Use this to tweak log, cache and concurrency settings.
   */
  options?: Partial<ScraperOptions>;
}

export interface ScraperOptions {
  /**
   * Enables writing pino logs to the stdout.
   * @default process.env.DEBUG === "true"
   */
  log: boolean;
  /**
   * Enables HTML payload caching on the disk.
   * Keep in mind that papercut **will not** clear the cache for you.
   * When enabling this, it's your responsability to deal with cache invalidation.
   *
   * @default false
   */
  cache: boolean;
  /**
   * Concurrency settings.
   */
  concurrency: {
    /**
     * Amount of concurrent promises for page scraping.
     * @default 2
     */
    page: number;
    /**
     * Amount of concurrent promises for node scraping.
     * @default 2
     */
    node: number;
    /**
     * Amount of concurrent promises for selector scraping.
     * @default 2
     */
    selector: number;
  };
}

export const defaultOptions: ScraperOptions = {
  log: process.env.DEBUG === 'true',
  cache: false,
  concurrency: {
    page: 2,
    node: 2,
    selector: 2,
  },
};

export type Scraper = ReturnType<typeof createScraper>;

/**
 * Creates a new scraper runner.
 *
 * This method is papercut entrypoint. It will create
 * an Scraper struct containing a runner that you can tweak
 * as needed.
 *
 * The runner is going to abide to the settings given
 * during the creation of this object.
 *
 * This function will also create a pino logger
 * and embed it within the runner.
 *
 * In case you prefer to manage the logger yourself,
 * please use `createRunner` instead.
 */
export const createScraper = (props: ScraperProps) => {
  const options: ScraperOptions = {
    ...defaultOptions,
    ...props.options,
    concurrency: {
      ...defaultOptions.concurrency,
      ...props.options?.concurrency,
    },
  };

  const logger = createLogger({
    name: props.name,
    enabled: options.log,
  });

  /**
   * The scraper struct.
   */
  const scraper = {
    run: createRunner({
      logger,
      options,
    }),
  };

  return scraper;
};
