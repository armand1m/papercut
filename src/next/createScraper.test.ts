import path from 'path';

import { createScraper } from './createScraper';
process.env.PAPERCUT_PAGE_CACHE_PATH = path.resolve(
  __dirname,
  './__fixtures__/pagecache'
);

test('createScraper', async () => {
  const scraper = createScraper({
    name: `Hacker News`,
    options: {
      log: false,
      cache: true,
    },
  });

  const results = await scraper.run({
    strict: true,
    baseUrl: 'https://news.ycombinator.com/',
    target: '.athing',
    selectors: {
      rank: (utils) => {
        const value = utils.text('.rank').replace(/^\D+/g, '');
        return Number(value);
      },
      name: ({ text }) => text('.titlelink'),
      url: ({ href }) => href('.titlelink'),
      score: ({ element }) => {
        return element.nextElementSibling?.querySelector('.score')
          ?.textContent;
      },
      createdBy: ({ element }) => {
        return element.nextElementSibling?.querySelector('.hnuser')
          ?.textContent;
      },
      createdAt: ({ element }) => {
        return element.nextElementSibling
          ?.querySelector('.age')
          ?.getAttribute('title');
      },
    },
  });

  expect(results).toMatchSnapshot();
});
