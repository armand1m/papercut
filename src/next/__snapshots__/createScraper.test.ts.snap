// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`createScraper 1`] = `
Array [
  Object {
    "createdAt": "2021-11-13T17:18:12",
    "createdBy": "picture",
    "name": "Reverse-engineering the Yamaha DX7 synthesizer's sound chip from die photos",
    "rank": 1,
    "score": "190 points",
    "url": "https://www.righto.com/2021/11/reverse-engineering-yamaha-dx7.html",
  },
  Object {
    "createdAt": "2021-11-13T13:04:37",
    "createdBy": "IceWreck",
    "name": "Restic – Backups Done Right",
    "rank": 2,
    "score": "301 points",
    "url": "https://restic.net/",
  },
  Object {
    "createdAt": "2021-11-13T15:00:26",
    "createdBy": "dinkelberg",
    "name": "Software company SAP: World-leading through intellectual property theft?",
    "rank": 3,
    "score": "175 points",
    "url": "https://www.tagesschau.de/investigativ/mdr/sap-geistiges-eigentum-101.html",
  },
  Object {
    "createdAt": "2021-11-13T09:13:49",
    "createdBy": "h1x",
    "name": "It's now possible to sign arbitrary data with your SSH keys",
    "rank": 4,
    "score": "480 points",
    "url": "https://www.agwa.name/blog/post/ssh_signatures",
  },
  Object {
    "createdAt": "2021-11-13T17:33:39",
    "createdBy": "tosh",
    "name": "Project Euler",
    "rank": 5,
    "score": "148 points",
    "url": "https://projecteuler.net",
  },
  Object {
    "createdAt": "2021-11-13T14:52:08",
    "createdBy": "krycroft",
    "name": "Show HN: Marvin – A grumpy Slackbot who doesn't like your ideas",
    "rank": 6,
    "score": "65 points",
    "url": "https://github.com/keltrycroft/marvin",
  },
  Object {
    "createdAt": "2021-11-13T12:39:45",
    "createdBy": "mooreds",
    "name": "Life is too short to depend on unstable software",
    "rank": 7,
    "score": "148 points",
    "url": "https://blog.sidebits.tech/life-is-too-short-to-depend-on-unstable-software/",
  },
  Object {
    "createdAt": "2021-11-13T16:09:40",
    "createdBy": "_Microft",
    "name": "Glass is the hidden gem in a carbon-neutral future",
    "rank": 8,
    "score": "88 points",
    "url": "https://www.nature.com/articles/d41586-021-02992-8",
  },
  Object {
    "createdAt": "2021-11-12T13:15:08",
    "createdBy": "nz",
    "name": "Gin, television, and social surplus, or, “looking for the mouse” (2008)",
    "rank": 9,
    "score": "63 points",
    "url": "https://gist.github.com/jm3/6724931",
  },
  Object {
    "createdAt": "2021-11-13T08:13:08",
    "createdBy": "jacksoncloud",
    "name": "Email from FBI Looks Odd",
    "rank": 10,
    "score": "373 points",
    "url": "https://old.reddit.com/r/sysadmin/comments/qsun7o/email_from_fbi_looks_odd/",
  },
  Object {
    "createdAt": "2021-11-13T14:49:12",
    "createdBy": "malshe",
    "name": "In the 17th Century, Leibniz Dreamed of a Machine That Could Calculate Ideas",
    "rank": 11,
    "score": "51 points",
    "url": "https://spectrum.ieee.org/in-the-17th-century-leibniz-dreamed-of-a-machine-that-could-calculate-ideas",
  },
  Object {
    "createdAt": "2021-11-13T16:27:53",
    "createdBy": "DerekBickerton",
    "name": "Japanese Philosophies That'll Help You Spend Money Consciously",
    "rank": 12,
    "score": "52 points",
    "url": "https://hulry.com/japanese-philosophies-money/",
  },
  Object {
    "createdAt": "2021-11-12T19:42:20",
    "createdBy": "brudgers",
    "name": "Minolta X1 (2007)",
    "rank": 13,
    "score": "34 points",
    "url": "http://www.rokkorfiles.com/XK.html",
  },
  Object {
    "createdAt": "2021-11-13T17:00:49",
    "createdBy": undefined,
    "name": "Shelf Engine (YC S18) Is Hiring a Data Science Manager. Help Us Stop Food Waste",
    "rank": 14,
    "score": undefined,
    "url": "https://www.shelfengine.com/careers/?gh_jid=4099404004",
  },
  Object {
    "createdAt": "2021-11-13T17:54:10",
    "createdBy": "sendilkumarn",
    "name": "Cloudflare blocks an almost 2 Tbps multi-vector DDoS attack",
    "rank": 15,
    "score": "92 points",
    "url": "https://blog.cloudflare.com/cloudflare-blocks-an-almost-2-tbps-multi-vector-ddos-attack/",
  },
  Object {
    "createdAt": "2021-11-13T16:13:39",
    "createdBy": "atlasunshrugged",
    "name": "Monopolies Slowly Destroying The Pentagon",
    "rank": 16,
    "score": "119 points",
    "url": "https://mattstoller.substack.com/p/monopolies-slowly-destroying-the",
  },
  Object {
    "createdAt": "2021-11-13T12:54:57",
    "createdBy": "Bostonian",
    "name": "The Breeder's Equation",
    "rank": 17,
    "score": "81 points",
    "url": "https://www.edge.org/response-detail/27199",
  },
  Object {
    "createdAt": "2021-11-13T03:53:59",
    "createdBy": "tjalfi",
    "name": "The weirdest bug I've ever encountered",
    "rank": 18,
    "score": "24 points",
    "url": "https://mental-reverb.com/blog.php?id=29",
  },
  Object {
    "createdAt": "2021-11-13T03:42:18",
    "createdBy": "zdw",
    "name": "Sound of the 2C33: The Famicom Disk System",
    "rank": 19,
    "score": "7 points",
    "url": "https://nicole.express/2021/music-for-the-whole-family.html",
  },
  Object {
    "createdAt": "2021-11-13T13:14:18",
    "createdBy": "theduder99",
    "name": "Chemists discover new way to harness energy from ammonia",
    "rank": 20,
    "score": "27 points",
    "url": "https://phys.org/news/2021-11-chemists-harness-energy-ammonia.html",
  },
  Object {
    "createdAt": "2021-11-13T18:05:15",
    "createdBy": "nixcraft",
    "name": "Dte: A language for expressing and calculating date and time",
    "rank": 21,
    "score": "13 points",
    "url": "https://github.com/mvrozanti/dte",
  },
  Object {
    "createdAt": "2021-11-13T15:29:19",
    "createdBy": "kxrm",
    "name": "Mailgun doesn't validate the email headers of email sent through their system",
    "rank": 22,
    "score": "103 points",
    "url": "https://twitter.com/RadioToolbox/status/1459542957275717633",
  },
  Object {
    "createdAt": "2021-11-13T19:31:29",
    "createdBy": "rectang",
    "name": "Would You Manage 70 Children and a 15-Ton Vehicle for $18 an Hour?",
    "rank": 23,
    "score": "29 points",
    "url": "https://fivethirtyeight.com/features/would-you-manage-70-children-and-a-15-ton-vehicle-for-18-an-hour/",
  },
  Object {
    "createdAt": "2021-11-13T15:38:14",
    "createdBy": "infodocket",
    "name": "Agriculture Gave Rise to One of World’s Most Mysterious Language Families",
    "rank": 24,
    "score": "34 points",
    "url": "https://www.science.org/content/article/how-agriculture-gave-rise-one-world-s-most-mysterious-language-families",
  },
  Object {
    "createdAt": "2021-11-12T17:13:58",
    "createdBy": "boulos",
    "name": "The Design of the Connection Machine (1994)",
    "rank": 25,
    "score": "75 points",
    "url": "https://tamikothiel.com/theory/cm_txts/",
  },
  Object {
    "createdAt": "2021-11-13T18:27:11",
    "createdBy": "atlasunshrugged",
    "name": "Nuclear Fusion Is Close Enough to Start Dreaming",
    "rank": 26,
    "score": "43 points",
    "url": "https://www.bloomberg.com/opinion/articles/2021-11-11/nuclear-fusion-is-close-enough-to-start-dreaming-of-a-new-world",
  },
  Object {
    "createdAt": "2021-11-13T18:55:22",
    "createdBy": "bryan0",
    "name": "Should California’s last nuclear power plant stay open?",
    "rank": 27,
    "score": "18 points",
    "url": "https://www.sandiegouniontribune.com/business/story/2021-11-12/should-californias-last-nuclear-power-plant-stay-open",
  },
  Object {
    "createdAt": "2021-11-13T17:03:47",
    "createdBy": "todsacerdoti",
    "name": "Managing an External Display on Linux Shouldn’t Be This Hard",
    "rank": 28,
    "score": "10 points",
    "url": "https://changelog.complete.org/archives/10311-managing-an-external-display-on-linux-shouldnt-be-this-hard",
  },
  Object {
    "createdAt": "2021-11-12T22:14:39",
    "createdBy": "strang1ato",
    "name": "Show HN: Nhi – capture every information about each executed command [video]",
    "rank": 29,
    "score": "27 points",
    "url": "https://www.youtube.com/watch?v=i7F3fJdYXSs",
  },
  Object {
    "createdAt": "2021-11-11T10:57:30",
    "createdBy": "Amorymeltzer",
    "name": "The Geometry of Grief",
    "rank": 30,
    "score": "19 points",
    "url": "https://www.themarginalian.org/2021/11/10/geometry-of-grief-michael-frame/",
  },
]
`;
